frame/support/src/traits.rs

```diff
@@ -347,68 +366,63 @@ impl<T> Happened<T> for () {
 /// be the default value), or where the account is being removed or reset back to the default value
 /// where previously it did exist (though may have been in a default state). This works well with
 /// system module's `CallOnCreatedAccount` and `CallKillAccount`.
-pub struct StorageMapShim<
-	S,
-	Created,
-	Removed,
-	K,
-	T
->(sp_std::marker::PhantomData<(S, Created, Removed, K, T)>);
+pub struct StorageMapShim<S, L, K, T>(sp_std::marker::PhantomData<(S, L, K, T)>);
```

Comment:
sidenote:

it is a bit confusing to me, like the `StoredMap` always say that they make no distinction between insert default value and removing the value, but this implementation makes a distrinction between the 2.

Thus if someone want to use `StoredMap`, they can make no distrinction between insert default or removing the value. On the other end if someone wants to use `StorageMapShim` as implementation of `StoredMap` then have to be careful that the trait usage correctly make a difference between inserting default and removing the value.

For instance pallet balances encourages to use `StorageMapShim` but never really tells that they make a distinction between inserting default and removing.

A newcomer to the codebase could modify pallet balances thinking they do not have to care about the difference between inserting default and removing. And they would break `StorageMapShim`.

At some point I would prefer to reduce the complexity and only keep the `*_exists` methods. And `frame_system::AccountInfo::data` would be an `Option`.

Small additional note: `StoredMap::mutate` would use `T::default` on no value. whereas `StorageMap::mutate` will use the value provided to the macro. I guess this is what is meant by `"however this would break the ability to have custom impls of StoredValue"` but ppl could find it difficult to grasp.

But anyway this is not an issue of this PR at all.

Reply:
Switching `data` to `Option` would be good, I agree. The only question is whether to do it in this PR.

Reply:
Indeed, better to have it in another PR, this PR makes complete sense by itself